nav2_behavior_tree/src/behavior_tree_engine.cpp

```diff
@@ -16,51 +16,20 @@
 
 #include <memory>
 #include <string>
+#include <vector>
 
-#include "nav2_behavior_tree/back_up_action.hpp"
-#include "nav2_behavior_tree/bt_conversions.hpp"
-#include "nav2_behavior_tree/compute_path_to_pose_action.hpp"
-#include "nav2_behavior_tree/follow_path_action.hpp"
-#include "nav2_behavior_tree/goal_reached_condition.hpp"
-#include "nav2_behavior_tree/is_stuck_condition.hpp"
-#include "nav2_behavior_tree/rate_controller_node.hpp"
-#include "nav2_behavior_tree/recovery_node.hpp"
-#include "nav2_behavior_tree/spin_action.hpp"
-#include "nav2_behavior_tree/wait_action.hpp"
-#include "nav2_behavior_tree/clear_costmap_service.hpp"
-#include "nav2_behavior_tree/reinitialize_global_localization_service.hpp"
 #include "rclcpp/rclcpp.hpp"
 
 using namespace std::chrono_literals;
 
 namespace nav2_behavior_tree
 {
 
-BehaviorTreeEngine::BehaviorTreeEngine()
+BehaviorTreeEngine::BehaviorTreeEngine(const std::vector<std::string> & plugin_libraries)
 {
-  // Register our custom action nodes so that they can be included in XML description
-  factory_.registerNodeType<nav2_behavior_tree::ComputePathToPoseAction>("ComputePathToPose");
-  factory_.registerNodeType<nav2_behavior_tree::FollowPathAction>("FollowPath");
-  factory_.registerNodeType<nav2_behavior_tree::BackUpAction>("BackUp");
-  factory_.registerNodeType<nav2_behavior_tree::SpinAction>("Spin");
-  factory_.registerNodeType<nav2_behavior_tree::WaitAction>("Wait");
-  factory_.registerNodeType<nav2_behavior_tree::ClearEntireCostmapService>("ClearEntireCostmap");
-  factory_.registerNodeType<nav2_behavior_tree::ReinitializeGlobalLocalizationService>(
-    "ReinitializeGlobalLocalization");
-
-  // Register our custom condition nodes
-  factory_.registerNodeType<nav2_behavior_tree::IsStuckCondition>("IsStuck");
-  factory_.registerNodeType<nav2_behavior_tree::GoalReachedCondition>("GoalReached");
-
-  // Register our simple condition nodes
-  factory_.registerSimpleCondition("initialPoseReceived",
-    std::bind(&BehaviorTreeEngine::initialPoseReceived, this, std::placeholders::_1));
-
-  // Register our custom decorator nodes
-  factory_.registerNodeType<nav2_behavior_tree::RateController>("RateController");
-
-  // Register our custom control nodes
-  factory_.registerNodeType<nav2_behavior_tree::RecoveryNode>("RecoveryNode");
+  for (const auto & p : plugin_libraries) {
+    factory_.registerFromPlugin(std::string{"lib" + p + ".so"});
```

Comment:
This isn't going to work for non-linux systems. We need to find a way to make this cross-platform.

Maybe you can take a look at the pluginlib/class loader code to see how they make it cross-platform

Reply:
Right, that seemed something to be addressed upstream.

In the meantime, we can try what [pluginlib does](https://github.com/ros/pluginlib/blob/37e1763edb988e3c4e07de76a4e22163a81b5219/pluginlib/include/pluginlib/class_loader_imp.hpp#L377) which is to get a list of all possible paths and filename combinations to try...